### ISSUE TITLE
A device with app protection policies and conditional access applied doesn't call the completion closure during single sign-on login to Box with Edge.

### ISSUE THREAD

**Issue by yamamichi-datt** [msintuneappsdk/ms-intune-app-sdk-ios#410]
Hello, development team.
Can anyone help me with the following issue?
If it is wrong to ask this question in this repository, please tell me which repository is more appropriate to ask the question, [MSAL](https://github.com/AzureAD/microsoft-authentication-library-for-objc) or [box-ios-sdk](https://github.com/box/box-ios-sdk).

**Describe the bug:**

When performing single sign-on (SSO) authentication to online storage (Box) via Edge using an iOS app configured for conditional access, an error -34018 (errSecMissingEntitlement) occurs inside MSAL and the login process stops.

**To Reproduce**

I created a sample app with minimal implementation.
The app has only a "Login" button, and tapping this button will attempt to log in to Box using OAuth 2.0 after checking the conditional policy.

1. Launch the app.
1. Tap the "Login" button.
1. Check for conditional access with `loginAndEnrollAccount(nil)`.
1. An alert message appears requesting a restart. After the alert is closed, the application terminates.
1. Launch the app again.
1. Tap the "Login" button.
1. The in-app browser (WKWebView) will open and the login form will appear.
1. Switch to SSO login from the link at the bottom of the login form.
1. Enter an organization's email address.
1. "You can't get there from here" and the "Launch in Edge" button appears.
1. Tap the "Launch in Edge" button.
1. Edge is automatically launched.
1. Sign in to the organization's account, if necessary, and tap the "Allow access to box" button.
1. The process is automatically returned to the app.
1. The problem occurs at this time. Multiple -34018 (errSecMissingEntitlement) errors are logged, the in-app browser is not closed and the completion closure to receive the result is never called.

**Expected behavior:**

In the last step, the authentication process is accepted by Box and the in-app browser is automatically closed.

**Screenshots and logs:**

- If applicable, add screenshots to help explain your problem.
  - Please see "Additional context" section.

- If your app is experiencing an issue during build, do you have app logs for the crash itself?
  - The app did not crash.

- If your app is experiencing an issue post build, do you have app logs for the crash itself?
  - Please see "Additional context" section.

**Smartphone (please complete the following information):**

 - Device: iPad Air (5th generation)
 - OS: iPadOS 17.2

**Intune App SDK for iOS (please complete the following information):**

- What version of the Intune SDK are you using? Are you using the latest version?
  - 19.2.0

- What platform is your app based in (native, Xamarin based, Cordova, etc)?
  - Native

- For errors during build, does the app build without Intune SDK integration?
  - No errors during build.

- For errors post build, does the app launch without being Intune SDK integrated?
  - Yes, it does.

- Who is the customer?
  - All App Store users, including non-Intune users.

- Do you see a trend with it only being reproduced on a specific device?
  - No.

**Additional context:**

### Keychain Access Groups

```
<key>keychain-access-groups</key>
<array>
    <string>$(AppIdentifierPrefix)jp.co.datt.SampleApp.Verification</string>
    <string>$(AppIdentifierPrefix)com.microsoft.adalcache</string>
    <string>$(AppIdentifierPrefix)com.microsoft.intune.mam</string>
</array>
```

### Info.plist

`ADALLogOverrideDisabled` is set to `true` as we need to check the logs for this defect and will be removed as soon as this issue is resolved.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>CFBundleURLTypes</key>
    <array>
        <dict>
            <key>CFBundleTypeRole</key>
            <string>Editor</string>
            <key>CFBundleURLName</key>
            <string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
                <string>$(PRODUCT_BUNDLE_IDENTIFIER)-intunemam</string>
                <string>msauth.$(PRODUCT_BUNDLE_IDENTIFIER)</string>
                <string>msauth.$(PRODUCT_BUNDLE_IDENTIFIER)-intunemam</string>
            </array>
        </dict>
        <dict>
            <key>CFBundleTypeRole</key>
            <string>Editor</string>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>boxsdk-$(BOX_SDK_CLIENT_ID)</string>
                <string>boxsdk-$(BOX_SDK_CLIENT_ID)-intunemam</string>
                <string>msauth.com.microsoft.intunemam</string>
            </array>
        </dict>
    </array>
    <key>IntuneMAMSettings</key>
    <dict>
        <key>ADALAuthority</key>
        <string>https://login.microsoftonline.com/common</string>
        <key>ADALClientId</key>
        <string>********</string>
        <key>ADALLogOverrideDisabled</key>
        <true/>
        <key>ADALRedirectUri</key>
        <string>msauth.$(PRODUCT_BUNDLE_IDENTIFIER)://auth</string>
    </dict>
    <key>LSApplicationQueriesSchemes</key>
    <array>
        <string>scmx</string>
        <string>smsec</string>
        <string>microsoft-edge-http</string>
        <string>microsoft-edge-http-intunemam</string>
        <string>lookoutwork-ase</string>
        <string>mvisionmobile</string>
        <string>skycure</string>
        <string>companyportal</string>
        <string>lacoonsecurity</string>
        <string>msauthv2</string>
        <string>wandera</string>
        <string>betteractiveshield</string>
        <string>msauthv3</string>
        <string>microsoft-edge-https</string>
        <string>microsoft-edge-https-intunemam</string>
        <string>zips</string>
        <string>ms-outlook</string>
        <string>ms-outlook-intunemam</string>
        <string>smart-ns</string>
        <string>https-intunemam</string>
        <string>http-intunemam</string>
        <string>intunemam-mtd</string>
    </array>
</dict>
</plist>
```

### API permissions (Entra)

- Configured permissions
  - Microsoft Graph
    - User.Read
  - Microsoft Mobile Application Management
    - DeviceManagementManagedApps.ReadWrite

#### Apps

- Target to apps on all device types
  - Yes
- Device types
  - No Device types
- Public apps
  - No Public apps
- Custom apps
  - jp.co.datt.sampleapp.verification

#### Data protection

- Prevent backups
  - Block
- Send org data to other apps
  - Policy managed apps
- Select apps to exempt
  - Default (I didn't change)
- Select universal links to exempt
  - Default (I didn't change)
- Select managed universal links
  - Default (I didn't change)
- Save copies of org data
  - Block
- Allow user to save copies to selected services
  - Box
- Transfer telecommunication data to
  - Any dialer app
- Dialer App URL Scheme
  - No Dialer App URL Scheme
- Transfer messaging data to
  - Any messaging app
- Messaging App URL Scheme
  - No Messaging App URL Scheme
- Receive data from other apps
  - All Apps
- Open data into Org documents
  - Allow
- Allow users to open data from selected services
  - OneDrive for Business
  - SharePoint
  - Camera
  - Photo Library
- Restrict cut, copy, and paste between other apps
  - Policy managed apps with paste in
- Cut and copy character limit for any app
  - 0
- Third party keyboards
  - Allow
- Encrypt org data
  - Require
- Sync policy managed app data with native apps or add-ins
  - Block
- Printing org data
  - Allow
- Restrict web content transfer with other apps
  - Microsoft Edge
- Unmanaged browser protocol
  - No Unmanaged browser protocol
- Org data notifications
  - Allow

#### Apps

- Target to apps on all device types
  - Yes
- Device types
  - No Device types
- Public apps
  - Microsoft Edge
- Custom apps
  - No Custom apps

#### Data protection

- Prevent backups
  - Block
- Send org data to other apps
  - All Apps
- Select apps to exempt
  - Default (I didn't change)
- Select universal links to exempt
  - Default (I didn't change)
- Select managed universal links
  - Default (I didn't change)
- Save copies of org data
  - Block
- Allow user to save copies to selected services
  - Box
  - Local Storage
  - OneDrive for Business
  - SharePoint
- Transfer telecommunication data to
  - Any dialer app
- Dialer App URL Scheme
  - No Dialer App URL Scheme
- Transfer messaging data to
  - Any messaging app
- Messaging App URL Scheme
  - No Messaging App URL Scheme
- Receive data from other apps
  - All Apps
- Open data into Org documents
  - Allow
- Allow users to open data from selected services
  - OneDrive for Business
  - SharePoint
  - Camera
  - Photo Library
- Restrict cut, copy, and paste between other apps
  - Policy managed apps with paste in
- Cut and copy character limit for any app
  - 0
- Third party keyboards
  - Allow
- Encrypt org data
  - Require
- Sync policy managed app data with native apps or add-ins
  - Block
- Printing org data
  - Allow
- Restrict web content transfer with other apps
  - Microsoft Edge
- Unmanaged browser protocol
  - No Unmanaged browser protocol
- Org data notifications
  - Allow

#### Conditions

- Device platforms
  - iOS

#### Access controls

- Grant
  - Grant access
    - Require device to be marked as compliant
    - Require app protection policy
  - For multiple controls
    - Require all the selected controls

### Logs

This is the log output after returning from Edge to the sample app.

```
TID=523199 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Complete cert auth challenge with end URL: boxsdk-********://boxsdkoauth2redirect?state=(not-null)&code=(not-null)
result: false
TID=523199 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] This broker response cannot be handled. Skipping request.
TID=523770 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Querying device info
TID=523770 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] GetDeviceInfo: Should call Sso Extension decision: 1
TID=523770 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] GetDeviceInfo: Creating Sso Extension request
TID=523770 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] GetDeviceInfo: Invoking Sso Extension with ssoExtensionRequest: Masked(not-null)
TID=523770 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] ASAuthorizationController: performRequests with operation(s): get_device_info
TID=523909 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Querying device info
TID=523909 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] GetDeviceInfo: Should call Sso Extension decision: 1
TID=523909 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] GetDeviceInfo: Creating Sso Extension request
TID=523909 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] GetDeviceInfo: Invoking Sso Extension with ssoExtensionRequest: Masked(not-null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default cache) Get app metadata entries with clientId ********, environment (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: appmetadata-******** generic: (null) type: 3001, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523909 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] ASAuthorizationController: performRequests with operation(s): get_device_info
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Get accounts.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Get accounts with environment (null), clientId ********, familyId , account (null), username (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default cache) Get accounts with environment (null), unique user id Masked(null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: (null) type: 1003, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieving cached credentials using credential query
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: (null) type: 2002, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: -25300
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 0 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieved 0 cached credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) credential query requires exact match with the cached credential items. Performing additional filtering checks.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) returning 0 filtered credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Looking for token with aliases (null), tenant (null), clientId ********, scopes (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieving cached credentials using credential query
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: (null) type: 2003, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieved 1 cached credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) credential query requires exact match with the cached credential items. Performing additional filtering checks.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) performing filtering check on cached credential item with the following properties - client ID: ********, target: (null), realm: eea9e461-64da-4d08-ad55-680923b7a47c, environment: login.windows.net, familyID: (null), homeAccountId: 9fd3a606, enrollmentId: Masked(null), appKey: Masked(null), applicationIdentifier: Masked(null), tokenType: (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) returning 1 filtered credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Found 1 tokens
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(not-null) service: ******** generic: (null) type: 5001, keychainGroup: adal keychain group)
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] GetDeviceInfo: Receiving results from Sso Extension with device info: Masked(not-null), error: Masked(null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] Checking Legacy keychain for registration.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Found the following accounts in default accessor: MaskedArray(count=1)
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Failed to find workplace join private key with status -34018
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Legacy accessor) Get accounts with environment (null), clientId ********, familyId , account identifier (null), legacy identifier (null)
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] Checking keychain for default registration done using ECC key.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: {length = 17, bytes = 0x4d534f70656e546563682e4144414c2e31} type: (null), keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Error Domain=MSIDKeychainErrorDomain Code=-34018 "(null)" UserInfo={MSIDCorrelationIdKey=7A5A8FD4-6711-4682-8E32-CC7DDB8E098E, MSIDErrorDescriptionKey=Could not get default primary registration tenantId.}
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] GetDeviceInfo: Completing filling device info: Masked(not-null), error: Masked(null)
TID=523477 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 7A5A8FD4-6711-4682-8E32-CC7DDB8E098E] [MSAL] Retrieved device info Masked(not-null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: -25300
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 0 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Legacy accessor) Found no refresh tokens
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to get wipe info...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Wipe query: Masked(not-null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get wipe info status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Found the following accounts in other accessors: MaskedArray(count=1)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default cache) Get app metadata entries with clientId ********, environment (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: appmetadata-******** generic: (null) type: 3001, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Get accounts.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Get accounts with environment (null), clientId ********, familyId , account (null), username (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default cache) Get accounts with environment (null), unique user id Masked(null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: (null) type: 1003, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieving cached credentials using credential query
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: (null) type: 2002, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: -25300
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 0 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieved 0 cached credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) credential query requires exact match with the cached credential items. Performing additional filtering checks.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) returning 0 filtered credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Looking for token with aliases (null), tenant (null), clientId ********, scopes (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieving cached credentials using credential query
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: (null) type: 2003, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) retrieved 1 cached credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) credential query requires exact match with the cached credential items. Performing additional filtering checks.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) performing filtering check on cached credential item with the following properties - client ID: ********, target: (null), realm: eea9e461-64da-4d08-ad55-680923b7a47c, environment: login.windows.net, familyID: (null), homeAccountId: 9fd3a606, enrollmentId: Masked(null), appKey: Masked(null), applicationIdentifier: Masked(null), tokenType: (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (MSIDAccountCredentialCache) returning 1 filtered credentials
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Found 1 tokens
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(not-null) service: ******** generic: (null) type: 5001, keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 1 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Found the following accounts in default accessor: MaskedArray(count=1)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Legacy accessor) Get accounts with environment (null), clientId ********, familyId , account identifier (null), legacy identifier (null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get keychain items, key info (account: Masked(null) service: (null) generic: {length = 17, bytes = 0x4d534f70656e546563682e4144414c2e31} type: (null), keychainGroup: adal keychain group)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to find keychain items...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Keychain find status: -25300
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Found 0 items.
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Legacy accessor) Found no refresh tokens
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Trying to get wipe info...
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Wipe query: Masked(not-null)
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Get wipe info status: 0
TID=523915 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] (Default accessor) Found the following accounts in other accessors: MaskedArray(count=1)
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] GetDeviceInfo: Receiving results from Sso Extension with device info: Masked(not-null), error: Masked(null)
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] Checking Legacy keychain for registration.
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Failed to find workplace join private key with status -34018
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] Checking keychain for default registration done using ECC key.
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] Error Domain=MSIDKeychainErrorDomain Code=-34018 "(null)" UserInfo={MSIDCorrelationIdKey=1D4DD13D-6480-4390-9539-B4B88E8B5C55, MSIDErrorDescriptionKey=Could not get default primary registration tenantId.}
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39] GetDeviceInfo: Completing filling device info: Masked(not-null), error: Masked(null)
TID=523924 MSAL 1.2.20 iOS 17.2 [2024-03-12 07:56:39 - 1D4DD13D-6480-4390-9539-B4B88E8B5C55] [MSAL] Retrieved device info Masked(not-null)
```

### Screen recording

![screen_capture](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/assets/163077017/d57acdab-7fef-4392-a8d6-0e7b7b99f103)

**Comment by pmod2** [comment]
Hi @yamamichi-datt, thankyou for reporting this issue. 
1. Is this not working for a single user or the pattern is uniform across all users?
2. Was the app working before and not working now or is this a new implementation?

Please check if your sample app has the callback handled in the AppDelegate or UISceneDelegate (whichever applicable based on your iOS version compatibility) as mentioned [in the doc here](https://github.com/AzureAD/microsoft-authentication-library-for-objc?tab=readme-ov-file#configuring-msal) 
If you support both UISceneDelegate and UIApplicationDelegate for compatibility with older iOS, MSAL callback would need to be placed into both files.

**Comment by yamamichi-datt** [comment]
Hello @pmod2, Thank you for your reply.  
I will reply to your confirmations.

> 1. Is this not working for a single user or the pattern is uniform across all users?

I have not tried this with multiple users from the same organization.  
This issue occurred with one iPad and one account loaned by our customer.  
To test this issue, I signed up for a free trial of Intune and Entra ID P2.  
There is only one account registered there and it has the same problem as our customer.

> 2. Was the app working before and not working now or is this a new implementation?

This is a new implementation.  
Originally, the app did not include MSAL or Intune, but I was considering integrating the Intune SDK at the request of our customer.  

> Please check if your sample app has the callback handled in the AppDelegate or UISceneDelegate (whichever applicable based on your iOS version compatibility) as mentioned in the doc here
> If you support both UISceneDelegate and UIApplicationDelegate for compatibility with older iOS, MSAL callback would need to be placed into both files.

I have checked with a breakpoint.

The app that will integrate the Intune SDK only has AppDelegate, and `application(_:open:options:)` is called when redirected from Edge.  
The argument `url` is the Box redirect URL with the query string appended.  
The argument `options` is an empty Dictionary.  
`MSALPublicClientApplication.handleMSALResponse(_:sourceApplication:)` returns `false`.

Also, the SampleApp I created to test this bug has a SceneDelegate that implements `scene(_:openURLContexts:)`.  
This is also called, and `MSALPublicClientApplication.handleMSALResponse(_:sourceApplication:)` returns `false`.  

Thank you for your support.

Yamamichi

**Comment by kaisong1990** [comment]
Hi @yamamichi-datt  in your Steps to repro step 7, who presented the in-app browser? Asking because it seems to be a system browser instead of WKWebView, and there is an Internet Explore icon from the navigation bar which does not seem to be provided by MSAL.

Also, I saw the log:

**Broker response cannot be handled** which is expected as Edge is not a broker on the device.

My expectation here is after sign in from Edge, the in-app popup should be dismissed automatically regardless of the navigation back to your sample app or not.

**Comment by yamamichi-datt** [comment]
Hello @kaisong1990, thank you for your reply.

> in your Steps to repro step 7, who presented the in-app browser?

The in-app browser is presented by box-ios-sdk.

The in-app browser is automatically opened when the method that performs OAuth2 authentication is called.

I apologize for the confusion. I provided incorrect information.  
I checked the source code for box-ios-sdk version 5.5.0 and found that `ASWebAuthenticationSession` is called, not `WKWebView`.

> Broker response cannot be handled which is expected as Edge is not a broker on the device.

Am I doing something wrong on Intune?

Thank you for your support.

Yamamichi

**Comment by kaisong1990** [comment]
Hi @yamamichi-datt, thanks for checking!

I don't think this is related to MSAL SDK as the ASWebAuthenticationSession is started by box-ios-sdk . The **MSALPublicClientApplication.handleMSALResponse(_:sourceApplication:)** returns false is expected as the redirect came from Edge browser app, not broker (Authenticator app is broker on iOS). MSAL just does a check in case the response is from broker but not impact the main flow. Logging such message does not mean the error for this case.
 
The real problem is why ASWebAuthenticationSession is not calling its completion after user finished login from Edge browser. Since MSAL is not involved in the flow, I would suggest the follow ups:
 
1. Box development team if they handle the ASWebAuthenticationSession completion correctly when app is in background.
2. If #1 is Yes, reach out to Apple by providing system diagnose logs to understand why ASWebAuthenticationSession completion handling is not triggered. This is a system component, and I don't think box has control over the triggering condition.
If Apple mentioned something related to our auth service redirection, please create an IcM against our eSTS service team by providing correlation_id and timestamp.

**Comment by yamamichi-datt** [comment]
Hello @kaisong1990, thank you for the information.

I will contact the Box development team and report back here if any progress is made.

Thank you for your support.

Yamamihci

**Comment by yamamichi-datt** [comment]
Hello, development team.

I am writing to report on the progress.

I have contacted the Box development team.
Authentication while the app is in the background was not supported, and therefore, it did not work.
They recommended that I do one of the following:

- Ask my customer's admin to turn off the exclusion of Safari on the devices in Intune.
- Use CCG (Client Credentials Grant) authentication instead of OAuth2 authentication.

I intend to work on these.

Thank you for your advice.
As my question has now been resolved, may I close this issue?

Best regards,

Yamamichi
